docs/proposals/20250513-propogate-taints.md

```diff
@@ -0,0 +1,455 @@
+This follows the convention established by `cluster.x-k8s.io/labels-from-machine`.
+
+The taints will be concatenated with `,` and the serialization will look like this given the 4 ways a taint can be specified:
+`cluster.x-k8s.io/taints-from-machine:<key1>=<value1>:<effect1>,<key2>=<value2>:,<key3>:<effect3>,<key4>`
+
+See the upstream [string implementation](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/taint.go) for more details.
+
+##### Changes to the Cluster topology, MachineSet and MachineDeployment controllers
+
+The Cluster's topology controller will properly calculate and in-place propagate the new introduced taint fields down to the ControlPlane object (if it complies with the optional contract part), MachineDeployment and/or MachinePool resources.
+
+The MachineDeployment and MachineSet controllers will watch their associated `MachineSpec` for updates to the `spec.taints` field and update their owned objects objects (MachineSet or Machine) to reflect the new values.
+
+This should be done similar to how the existing in-place mutable fields like `ReadinessGates`, `NodeDrainTimeout`, etc. for a Machine are handled today.
+
+### Security Model
+
+Users who can define `Taint`s that get placed on `Node`s will be able to steer workloads, possibly to malicious hosts in order to extract sensitive data.
+However, users who can define Cluster API resources already have this capability - an attacker who receives the permissions to update a `MachineTemplate` could alter the definition in a similar manner.
+
+This proposal therefore does not present any heightened security requirements than Cluster API already has.
+
+#### Migration strategy for Bootstrap providers
+
+With the introduction of the new fields, bootstrap providers could start deprecating their equivalent fields.
+
+For CABPK, adding a taint at a KubeadmConfigTemplate at `.spec.template.spec.joinConfiguration.nodeRegistration.taints`. is almost equivalent to adding a `Initialization` typed taint by the new API.
+The only difference is that the taint does get added by the controllers after the Node joined the Cluster.
+
+This is considered okay, because workload should not be able to be schedule workload unless the `node.cluster.x-k8s.io/uninitialized` taint was removed and the implementation should take care to have added the `Initialization` taints before or at the time removing this taint.
+
+### Risks and Mitigations
+
+Managing the `Taint`s on `Node`s is considered a highly privileged action in Kubernetes; it even has its own top level `kubectl taint` command.
+The Kubernetes scheduler uses taints rather than `Conditions` to decide when to evict workloads.
+Updating these in place could then evict workloads unintentionally, or disrupt other systems that rely on taints being present.
+
+This risk can be mitigated by ensuring Cluster API only modifies taints that it owns on nodes, as decribed in the above.
+
```

Comment:
I don't remember which section this belongs to, can we consider feature gating this feature? This would give us some time to get more experience with this feature and make small adjustments if necessary